internal/search/query/literal_parser.go

```
@@ -0,0 +1,281 @@
+	}
+	if !p.expect(AND) {
+		return left, nil
+	}
+	right, err := p.parseAndLiteral()
+	if err != nil {
+		return nil, err
+	}
+	return newOperator(append(left, right...), And), nil
+}
+
+// parseOr parses or-expressions. Or operators have lower precedence than And
+// operators, therefore this function calls parseAnd.
+func (p *parser) parseOrLiteral() ([]Node, error) {
+	left, err := p.parseAndLiteral()
+	if err != nil {
+		return nil, err
+	}
+	if left == nil {
+		return nil, &ExpectedOperand{Msg: fmt.Sprintf("expected operand at %d", p.pos)}
+	}
+	if !p.expect(OR) {
+		return left, nil
+	}
+	right, err := p.parseOrLiteral()
+	if err != nil {
+		return nil, err
+	}
+	return newOperator(append(left, right...), Or), nil
+}
+
+func literalFallbackParser(in string) ([]Node, error) {
+	parser := &parser{
+		buf:        []byte(in),
+		heuristics: allowDanglingParens,
+	}
+	nodes, err := parser.parseOrLiteral()
+	if err != nil {
+		return nil, err
+	}
+	if hoistedNodes, err := Hoist(nodes); err == nil {
+		return newOperator(hoistedNodes, And), nil
+	}
+	return newOperator(nodes, And), nil
+}
+
+func ParseAndOrLiteral(in string) ([]Node, error) {
+	if strings.TrimSpace(in) == "" {
+		return nil, nil
+	}
+	parser := &parser{buf: []byte(in)}
+	nodes, err := parser.parseOrLiteral()
+	if err != nil {
+		switch err.(type) {
+		case *ExpectedOperand:
+			// The query may be unbalanced or malformed as in "(" or
+			// "x or" and expects an operand. Try harder to parse it.
+			if nodes, err := literalFallbackParser(in); err == nil {
+				return nodes, nil
+			}
+		}
+		// Another kind of error, like a malformed parameter.
+		return nil, err
+	}
+	if parser.balanced != 0 {
+		// The query is unbalanced and might be something like "(x" or
+		// "x or (x" where patterns start with a leading open
+		// parenthesis. Try harder to parse it.
+		if nodes, err := literalFallbackParser(in); err == nil {
+			return nodes, nil
+		}
+		return nil, errors.New("unbalanced expression")
```

The actual error is swallowed, maybe 
```suggestion
		return nil, errors.Wrap(err, "unbalanced expression")
```

---

I was OK with it being swallowed, but wrapping is better :-)